### PR TITLE
Closes race condition remaining from BVC-71

### PR DESCRIPTION
**Pull request by maxneuvians**
Refactors #159.

It looks like wrapping the SELECT in a transaction and locking it does not actually fix the race condition, just makes it less likely. This change should fix the race condition because it does not do a SELECT before an INSERT, but rather relies on the UNIQUE attribute of the column to handle the UNIQUEness factor. 

Adds a new `persistEncryptionKeyWithHashID` query function that handles the DELETE logic if a hashID has not been claimed yet. Also add two migrations, one ensuring backwards capability with existing keys and the other ensuring UNIQUEness on the hash_id column.

Two things learned:

- Transactions do not prevent race conditions
- NULL is not globally unique in MySQL rows

### CODE REVIEW

File: pkg/persistence/db.go [hunk 1]
```diff
@@ -163,12 +163,17 @@ func (c *conn) NewKeyClaim(region, originator, hashID string) (string, error) {
 		}
 
 		oneTimeCode := fmt.Sprintf("%08d", n)
-
-		err = persistEncryptionKey(c.db, region, originator, hashID, pub, priv, oneTimeCode)
+		if len(hashID) == 128 {
+			err = persistEncryptionKeyWithHashID(c.db, region, originator, hashID, pub, priv, oneTimeCode)
+		} else {
+			err = persistEncryptionKey(c.db, region, originator, pub, priv, oneTimeCode)
```
Comment (obrien-j) [inline pkg/persistence/db.go:169]: Curious if there isn't any reason why we couldn't just simplify this, or the parent call slightly, and generate our own hashID if there isn't one being passed in by the caller, and then return it.

From what I gather, we're going to need this pattern for the portal team anyway, and honestly, it's how I would have preferred we rolled with the initial integration.

File: pkg/persistence/db.go [hunk 2]
```diff
@@ -163,12 +163,17 @@ func (c *conn) NewKeyClaim(region, originator, hashID string) (string, error) {
...
+		}
 		if err == nil {
 			return oneTimeCode, nil
 		} else if strings.Contains(err.Error(), "used hashID found") {
 			return "", ErrHashIDClaimed
+		} else if strings.Contains(err.Error(), "regenerate hashID") {
+			log(nil, err).Warn("regenerating hashID")
```
Comment (obrien-j) [inline pkg/persistence/db.go:176]: should this read regenerating OTC?

File: pkg/persistence/db.go (hunk 1 above)
Comment (maxneuvians) [reply pkg/persistence/db.go:169]: We could - it puts the burden of storing the HashID on the client end though if it is not pre-determined and you would have to build another end point to validate if the OTC has been used.

File: pkg/persistence/db.go (hunk 2 above)
Comment (maxneuvians) [reply pkg/persistence/db.go]: Yes - fixing

File: pkg/persistence/db.go (hunk 1 above)
Comment (obrien-j) [reply pkg/persistence/db.go:169]: I was thinking just mimicing the behaviour of the withHashID path...

keyclaim -> returns OTC||randomHashIDweGenerate
keyclaim/{hashid} -> does what ever it does now.

Only distinctinction being, we nudge people into only ever calling keyclaim bare path once.

**Comment by maxneuvians** [comment]
This migration will conflict with #197  and will need to be resolved depending on which merges first.

**Pull request merged**